plugin/storage/integration/grpc_test.go

```diff
@@ -16,94 +16,37 @@
 package integration
 
 import (
-	"net"
 	"os"
 	"path"
-	"sync"
 	"testing"
 
 	"github.com/stretchr/testify/require"
 	"go.uber.org/zap"
-	googleGRPC "google.golang.org/grpc"
 
 	"github.com/jaegertracing/jaeger/pkg/config"
 	"github.com/jaegertracing/jaeger/pkg/metrics"
 	"github.com/jaegertracing/jaeger/pkg/testutils"
 	"github.com/jaegertracing/jaeger/plugin/storage/grpc"
-	grpcMemory "github.com/jaegertracing/jaeger/plugin/storage/grpc/memory"
-	"github.com/jaegertracing/jaeger/plugin/storage/grpc/shared"
-	"github.com/jaegertracing/jaeger/plugin/storage/memory"
 )
 
 const (
 	defaultPluginBinaryPath   = "../../../examples/memstore-plugin/memstore-plugin"
 	streamingPluginConfigPath = "fixtures/grpc_plugin_conf.yaml"
 )
 
-type gRPCServer struct {
-	errChan chan error
-	server  *googleGRPC.Server
-	wg      sync.WaitGroup
-}
-
-func newgRPCServer() (*gRPCServer, error) {
-	return &gRPCServer{errChan: make(chan error, 1)}, nil
-}
-
-func (s *gRPCServer) Restart() error {
-	// stop the server if one already exists
-	if s.server != nil {
-		s.server.GracefulStop()
-		s.wg.Wait()
-		select {
-		case err := <-s.errChan:
-			return err
-		default:
-		}
-	}
-
-	memStorePlugin := grpcMemory.NewStoragePlugin(memory.NewStore(), memory.NewStore())
-
-	s.server = googleGRPC.NewServer()
-	queryPlugin := shared.StorageGRPCPlugin{
-		Impl:        memStorePlugin,
-		ArchiveImpl: memStorePlugin,
-	}
-
-	if err := queryPlugin.RegisterHandlers(s.server); err != nil {
-		return err
-	}
-
-	listener, err := net.Listen("tcp", "localhost:2001")
-	if err != nil {
-		return err
-	}
-	s.wg.Add(1)
-	go func() {
-		defer s.wg.Done()
-		if err = s.server.Serve(listener); err != nil {
-			select {
-			case s.errChan <- err:
-			default:
-			}
-		}
-	}()
-	return nil
-}
-
 type GRPCStorageIntegrationTestSuite struct {
 	StorageIntegration
 	logger  *zap.Logger
 	flags   []string
 	factory *grpc.Factory
-	server  *gRPCServer
+	server  *GRPCServer
 }
 
 func (s *GRPCStorageIntegrationTestSuite) initialize(t *testing.T) {
 	s.logger, _ = testutils.NewLogger()
 
 	if s.server != nil {
-		err := s.server.Restart()
+		err := s.server.Start()
 		require.NoError(t, err)
 	}
```

Comment:
This doesn't look equivalent. The Restart function was stopping and replacing the server, you're starting the server that was already stopped (not even sure if it's safe to do)

Reply:
I believe the Restart function is an idempotent version of Start & Close functions.

When the Restart function is first called before running all the tests, it just starts the server. And then the Restart function is called in cleanUp function, which I've looked that it is called right before the end of every subtest, it stops the server and then replaces it with a new one.

With the new Start and Close functions, I "Start" the server before running the tests. Then in the cleanUp function, I "Close" the server and then "Start" a new one. After that, at the end of all tests, I "Close" the server (which the Restart function cannot do and causes goroutine leak).

By the way, the `if s.server != nil` is a check if the current test suite uses remote storage or a plugin (if it uses plugin, we don't start (or restart) the server. But, I'll add additional checking to the Start function to Close it first if it is already running.

Reply:
Could you please point out where the replacing behavior is? I'm afraid I missed your point.

Reply:
Old line 67
```
s.server = googleGRPC.NewServer()
```

Reply:
Oh, the Start function actually also replaces the server too. But I've refactored it so that the Start function is identical to Restart, with the extracted closing server logic into the Close function (so I can close the server after the end of all tests).